cdc/sinkv2/eventsink/mq/producer/kafka_producer.go

```diff
@@ -0,0 +1,241 @@
+	k.closedMu.RLock()
+	defer k.closedMu.RUnlock()
+
+	// If the producer is closed, we should skip the message.
+	if k.closed {
+		// We return the context error directly rather than other error.
+		// Because if producer already closed, it means the context maybe is canceling or canceled.
+		// So we shouldn't return other errors to the caller to mess up the shutdown process.
+		return ctx.Err()
+	}
+	failpoint.Inject("KafkaSinkAsyncSendError", func() {
+		// simulate sending message to input channel successfully but flushing
+		// message to Kafka meets error
+		log.Info("failpoint error injected", zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		k.failpointCh <- errors.New("kafka sink injected error")
+		failpoint.Return(nil)
+	})
+
+	failpoint.Inject("SinkFlushDMLPanic", func() {
+		time.Sleep(time.Second)
+		log.Panic("SinkFlushDMLPanic",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+	})
+
+	msg := &sarama.ProducerMessage{
+		Topic:     topic,
+		Partition: partition,
+		Key:       sarama.StringEncoder(message.Key),
+		Value:     sarama.ByteEncoder(message.Value),
+		Metadata:  messageMetaData{callback: message.Callback},
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.asyncProducer.Input() <- msg:
+	}
+	return nil
+}
+
+func (k *kafkaProducer) Close() error {
+	// We have to hold the lock to prevent write to closed producer.
+	k.closedMu.Lock()
+	defer k.closedMu.Unlock()
+	// If the producer was already closed, we should skip the close operation.
+	if k.closed {
+		// We need to guard against double closed the clients,
+		// which could lead to panic.
+		log.Warn("kafka producer already closed",
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID),
+			zap.Any("role", k.role))
+		return nil
+	}
+	close(k.failpointCh)
+	// Notify the run loop to exit.
+	close(k.closedChan)
+	k.closed = true
+	// `client` is mainly used by `asyncProducer` to fetch metadata and other related
+	// operations. When we close the `kafkaSaramaProducer`, TiCDC no need to make sure
+	// that buffered messages flushed.
+	// Consider the situation that the broker does not respond, If the client is not
+	// closed, `asyncProducer.Close()` would waste a mount of time to try flush all messages.
+	// To prevent the scenario mentioned above, close client first.
+	start := time.Now()
+	if err := k.client.Close(); err != nil {
+		log.Error("close sarama client with error", zap.Error(err),
+			zap.Duration("duration", time.Since(start)),
+			zap.String("namespace", k.id.Namespace),
+			zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+		return err
+	}
+	log.Info("sarama client closed", zap.Duration("duration", time.Since(start)),
+		zap.String("namespace", k.id.Namespace),
+		zap.String("changefeed", k.id.ID), zap.Any("role", k.role))
+
+	start = time.Now()
+	err := k.asyncProducer.Close()
```

Comment:
I'm not sure whether `k.client.Close` or `k.asyncProducer.Close()` could block for a long time, if it could, will the hold of `k.closedMu` cause some deadlock

Reply:
Yes, it will. So, how did the old one solve this problem?

Reply:
Perhaps it is acceptable that it causes the AsyncSendMessage to get stuck, since it does not affect the rest of the main goroutine.

Reply:
I will try to make it asynchronous in here, and I think that way we can solve the problem.

Reply:
> Yes, it will. So, how did the old one solve this problem?

Seems the old version has the same issue. I'm not sure whether we have met such deadlock problem.

Reply:
Maybe we can leave enough comments and a TODO here, and observe whether we can meet such problem in the future tests.

Reply:
Please checkout https://github.com/pingcap/tiflow/pull/6412/commits/2b637aab0945afded7aecafcb547da7ffb0d8778.